packages/backend/src/open_payments/payment/outgoing/service.ts

```diff
@@ -159,6 +185,326 @@ async function createOutgoingPayment(
+    BigInt(0)
+  )
+
+  const estTotalAvailableSendAmount =
+    availableSendAmount +
+    BigInt(
+      Math.floor(
+        (Number(payment.quote.lowEstimatedExchangeRate.b) /
+          Number(payment.quote.lowEstimatedExchangeRate.a)) *
+          (1 + slippage) *
+          Number(availableReceiveAmount)
+      )
+    )
+  const estTotalAvailableReceiveAmount =
+    availableReceiveAmount +
+    BigInt(
+      Math.floor(
+        (Number(payment.quote.lowEstimatedExchangeRate.a) /
+          Number(payment.quote.lowEstimatedExchangeRate.b)) *
+          (1 + slippage) *
+          Number(availableSendAmount)
+      )
+    )
+  return { estTotalAvailableSendAmount, estTotalAvailableReceiveAmount }
+}
+
+interface PaymentAccess extends GrantAccess {
+  paymentInterval?: Interval
+}
+
+// "payment" is locked by the "deps.knex" transaction.
+async function validateGrant(
+  deps: ServiceDependencies,
+  payment: OutgoingPayment,
+  grant: Grant
+): Promise<boolean> {
+  const grantAccess = grant.getAccess({
+    type: AccessType.OutgoingPayment,
+    action: AccessAction.Create
+  })
+
+  if (!grantAccess) {
+    // log?
+    return false
+  }
+
+  // Find grant access(es) that authorize this payment (pending send/receive limits).
+  const paymentAccess: PaymentAccess[] = []
+  const unlimitedAccess: GrantAccess[] = []
+  const unrelatedAccess: GrantAccess[] = []
+
+  for (const access of grantAccess) {
+    let noSendAmount = false
+    let noReceiveAmount = false
+    if (
+      validateAccess({
+        access,
+        payment,
+        identifier: `${deps.publicHost}/${payment.accountId}`
+      })
+    ) {
+      if (!access.limits) {
+        return true
+      }
+      if (!access.limits.sendAmount) {
+        noSendAmount = true
+      }
+      if (!access.limits.receiveAmount) {
+        noReceiveAmount = true
+      }
+      if (noSendAmount && noReceiveAmount) {
+        return true
+      }
+      let paymentInterval: Interval | undefined
+      if (access.interval) {
+        paymentInterval = getInterval(access.interval, payment.createdAt)
+        assert.ok(paymentInterval)
+      }
+      paymentAccess.push({
+        ...access,
+        paymentInterval
+      })
+    } else {
+      if (access.limits) {
+        unrelatedAccess.push(access)
+      } else {
+        unlimitedAccess.push(access)
+      }
+    }
+  }
+
+  if (paymentAccess.length === 0) {
+    return false
+  }
+  let {
+    estTotalAvailableSendAmount,
+    estTotalAvailableReceiveAmount
+  } = estimateTotalAvailableAmounts<PaymentAccess>(
+    paymentAccess,
+    payment,
+    deps.slippage
+  )
+
+  if (
+    estTotalAvailableSendAmount < payment.sendAmount.value ||
+    estTotalAvailableReceiveAmount < payment.receiveAmount.value
+  ) {
+    // Payment amount single-handedly exceeds sendAmount limit(s)
+    return false
+  }
+
+  // TODO: lock to prevent multiple grant payments from being created concurrently?
```

Comment:
bump

Reply:
Just to make sure I understand correctly, you want to lock querying of the grant payments to prevent another outgoing payment from being created concurrently?

Reply:
Yeah, I'm not sure if it's possible to just do an `insert` lock.

Reply:
When using `knex.raw`, [we can add a lock](https://stackoverflow.com/questions/52325771/manually-lock-table-in-knex). But I don't know if and how that works with the transaction.

Reply:
It'd be great if we could limit the lock to prevent inserts with the same `grantId`